matchzoo/metrics.py

```diff
@@ -1,190 +1,254 @@
+            >>> Precision(k=2)(y_true, y_pred)
+            0.0
+            >>> Precision(k=4)(y_true, y_pred)
+            0.25
+            >>> Precision(k=5)(y_true, y_pred)
+            0.2
+
+        :param y_true: The ground true label of each document.
+        :param y_pred: The predicted scores of each document.
+        :return: Precision @ k
+        :raises: ValueError: len(r) must be >= k.
+        """
+
+        if self._k <= 0:
+            raise ValueError('self._k must be larger than 0.')
+        coupled_pair = sort_and_couple(y_true, y_pred)
+        precision = 0.0
+        for idx, (label, score) in enumerate(coupled_pair):
+            if idx >= self._k:
+                break
+            if label > self._threshold:
+                precision += 1.
+        return precision / self._k
+
+
+class AveragePrecision(engine.BaseMetric):
+    ALIAS = 'average_precision'
+
+    def __init__(self, threshold=0):
+        self._threshold = threshold
+
+    def __repr__(self):
+        return f"{self.ALIAS}({self._threshold})"
+
+    def __call__(self, y_true, y_pred):
+        """
+        Calculate average precision (area under PR curve).
+
+        Example:
+            >>> y_true = [0, 1]
+            >>> y_pred = [0.1, 0.6]
+            >>> round(AveragePrecision()(y_true, y_pred), 2)
+            0.75
+
+        :param y_true: The ground true label of each document.
+        :param y_pred: The predicted scores of each document.
+        :return: Average precision.
+        """
+        precision_metrics = [Precision(k + 1) for k in range(len(y_pred))]
+        out = [metric(y_true, y_pred) for metric in precision_metrics]
+        if not out:
+            return 0.
+        return np.mean(out)
+
+
+class MeanAveragePrecision(engine.BaseMetric):
+    ALIAS = ['mean_average_precision', 'map', 'MAP']
+
+    def __init__(self, threshold=0):
+        self._threshold = threshold
+
+    def __repr__(self):
+        return f"{self.ALIAS[0]}({self._threshold})"
+
+    def __call__(self, y_true, y_pred):
+        """
+        Calculate mean average precision.
+
+        Example:
+            >>> y_true = [0, 1, 0, 0]
+            >>> y_pred = [0.1, 0.6, 0.2, 0.3]
+            >>> MeanAveragePrecision()(y_true, y_pred)
+            1.0
+
+        :param y_true: The ground true label of each document.
+        :param y_pred: The predicted scores of each document.
+        :return: Mean average precision.
+        """
+        result = 0.
+        pos = 0
+        coupled_pair = sort_and_couple(y_true, y_pred)
+        for idx, (label, score) in enumerate(coupled_pair):
+            if label > self._threshold:
+                pos += 1.
+                result += pos / (idx + 1.)
+        if pos == 0:
+            return 0.
+        else:
+            return result / pos
+
+
+class DiscountedCumulativeGain(engine.BaseMetric):
+    ALIAS = ['discounted_cumulative_gain', 'dcg', 'DCG']
+
+    def __init__(self, k=1, threshold=0):
```

Comment:
change the default `k` to 10? also, `k = 10` and `threshold = 0`
`pip install flake8 --upgrade`

Reply:
@faneshion had 1 but I don't know why.